robot_smach_states/src/robot_smach_states/manipulation/place_designator.py

```diff
@@ -0,0 +1,247 @@
+
+
+class EmptySpotDesignator(Designator):
+    """Designates an empty spot on the empty placement-shelve.
+    It does this by querying ED for entities that occupy some space.
+        If the result is no entities, then we found an open spot.
+
+    To test this in the robotics_test_lab with amigo-console:
+    robot = amigo
+    CABINET = "bookcase"
+    PLACE_SHELF = "shelf2"
+    cabinet = ds.EntityByIdDesignator(robot, id=CABINET, name="pick_shelf")
+    arm = ds.UnoccupiedArmDesignator(robot, {})
+    place_position = ds.LockingDesignator(ds.EmptySpotDesignator(robot, cabinet, arm, name="placement", area=PLACE_SHELF),
+                                          name="place_position")
+    """
+
+    def __init__(self, robot, place_location_designator, arm_designator, name=None, area=None):
+        """
+        Designate an empty spot (as PoseStamped) on some designated entity
+        :param robot: Robot whose worldmodel to use
+        :param place_location_designator: Designator resolving to an Entity, e.g. EntityByIdDesignator
+        :param arm_designator: Designator resolving to an arm robot part, e.g. OccupiedArmDesignator
+        :param name: name for introspection purposes
+        :param area: (optional) area where the item should be placed
+        """
+        super(EmptySpotDesignator, self).__init__(resolve_type=FrameStamped, name=name)
+        self.robot = robot
+
+        self.place_location_designator = place_location_designator
+        self.arm_designator = arm_designator
+        self._edge_distance = 0.05  # Distance to table edge
+        self._spacing = 0.15
+        self._area = area
+
+        self.marker_pub = rospy.Publisher('/empty_spots', MarkerArray, queue_size=1)
+        self.marker_array = MarkerArray()
+
+    def _resolve(self):
+        """
+        :return: Where can an object be placed
+        :returns: FrameStamped
+        """
+        place_location = self.place_location_designator.resolve()
+        place_frame = FrameStamped(frame=place_location._pose, frame_id="/map")
+
+        # points_of_interest = []
+        if self._area:
+            vectors_of_interest = self.determine_points_of_interest_with_area(place_location, self._area)
+        else:
+            vectors_of_interest = self.determine_points_of_interest(place_frame.frame, z_max=place_location.shape.z_max,
+                                                                    convex_hull=place_location.shape.convex_hull)
+
+        assert all(isinstance(v, FrameStamped) for v in vectors_of_interest)
+
+        open_POIs = filter(self.is_poi_occupied, vectors_of_interest)
+
+        base_pose = self.robot.base.get_location()
+        arm = self.arm_designator.resolve()
+        open_POIs_dist = [(poi, self.distance_to_poi_area_heuristic(poi, base_pose, arm)) for poi in open_POIs]
+
+        # We don't care about small differences
+        nav_threshold = 0.5 # Distance (0.5 m)
```

Comment:
Should we do something about this magic number? Maybe make it configurable with a default parameter?